proxy/marshal_test.go

```diff
@@ -0,0 +1,191 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proxy_test
+
+import (
+	"encoding/base64"
+	"strings"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/stretchr/testify/require"
+	"go.temporal.io/api/common/v1"
+	"go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/failure/v1"
+	"go.temporal.io/api/proxy"
+)
+
+var marshaler, _ = proxy.NewJSONPBMarshaler(proxy.JSONPBMarshalerOptions{})
+var unmarshaler, _ = proxy.NewJSONPBUnmarshaler(proxy.JSONPBUnmarshalerOptions{})
+
+func TestShorthandPayloads(t *testing.T) {
+	// Make a bunch of memo values to confirm payloads of different types
+	var memo common.Memo
+	binaryVal := base64.StdEncoding.EncodeToString([]byte("bytes"))
+	assertProtoJSON(t, &memo, `{
+		"fields": {
+			"some-string": "string",
+			"some-array": ["foo", 123, false],
+			"some-bool": false,
+			"some-float": 1.23,
+			"some-int": 123,
+			"some-null": null,
+			"some-bytes": {
+				"metadata": { "encoding": "`+base64.StdEncoding.EncodeToString([]byte("binary/plain"))+`" },
+				"data": "`+binaryVal+`"
+			},
+			"some-proto": {
+				"_protoMessageType": "temporal.api.common.v1.WorkflowType",
+				"name": "workflow-name"
+			},
+			"some-raw-payload": {
+				"metadata": { "encoding": "`+base64.StdEncoding.EncodeToString([]byte("my-encoding"))+`" },
+				"data": "`+base64.StdEncoding.EncodeToString([]byte("raw-value"))+`"
+			}
+		}
+	}`)
+	// Confirm some payloads
+	assertPayloadData(t, memo.Fields["some-string"], "json/plain", `"string"`)
+	assertPayloadData(t, memo.Fields["some-float"], "json/plain", "1.23")
+	assertPayloadData(t, memo.Fields["some-null"], "binary/null", "")
+	assertPayloadData(t, memo.Fields["some-bytes"], "binary/plain", "bytes")
+	assertPayloadJSONData(t, memo.Fields["some-proto"], "json/protobuf", `{"name":"workflow-name"}`)
+	require.Equal(t, "temporal.api.common.v1.WorkflowType", string(memo.Fields["some-proto"].Metadata["messageType"]))
+	assertPayloadData(t, memo.Fields["some-raw-payload"], "my-encoding", "raw-value")
+}
+
+func TestShorthandPayloadAmbiguity(t *testing.T) {
```

Comment:
@josh-berry - you might find this test valuable to read

Reply:
Got it, makes sense, thanks!